maintainers/README.md: expand section on maintainer tooling (#419690)
NixOS/nixpkgs: maintainers/README.md

The maintainer is welcome to come back at any time.

## Tools for maintainers

When a pull request is made against a package, nixpkgs CI will notify the

> appropriate maintainer(s) based on the package names listed in the commit titles.

Reviewer: I don't think it's based on commit messages — OfBorg builds are, but notifying maintainers is based on ci/OWNERS and looking at meta.maintainers for derivations that change.  It would be good to also clarify that submitters should still actively try to notify the appropriate people, because e.g. if they're just changing metadata for a package that doesn't result in a rebuild, there's a good chance CI won't be able to tell who to notify.

Author: That explains the new `meta.maintainersPosition`. However it is not perfect just like its brother `meta.position`, and even have different overriding semantics. more context in https://github.com/NixOS/nixpkgs/pull/359191#issuecomment-2982577523

Maintainers can also invoke the [nixpkgs-merge-bot](https://github.com/nixos/nixpkgs-merge-bot)
to merge pull requests targeting packages they are the maintainer of which satisfy the current
security [constraints](https://github.com/NixOS/nixpkgs-merge-bot/blob/main/README.md#constraints).

New maintainers will automatically get invited to join the [NixOS/nixpkgs-maintainers](https://github.com/orgs/NixOS/teams/nixpkgs-maintainers) GitHub team.
By joining it you will get some triaging rights in the nixpkgs repository, like the ability to add and remove labels.
However, the GitHub team invite is usually only sent by email and is only valid for one week!
Should it expire please ask for a re-invite in the [GitHub org owners help desk Matrix channel](https://matrix.to/#/#org_owners:nixos.org).

> [zh.fail](https://zh.fail/failed/overview.html) tracks all package build failures on `master` grouped by maintainer.

Reviewer 2: This is also quite unofficial I think

Author: Considering it being linked during zhf by the release managers i thought it was official :sweat_smile:

Reviewer 3: I was not aware of that link, thanks !

Reviewer 4: No, I hacked that together while i was release manager

Reviewer 5: Similar fail reports (no graphs, multiple branches) is at https://malob.github.io/nix-review-tools-reports/
That's especially useful for `staging-next*`, relatively easily estimating mass regressions.

Author: This is the resource is what i prefer to use during ZHF, but it does not have a per-maintainer view like zh.fail does

#### Unofficial maintainer tooling

[asymmetric/nixpkgs-update-notifier](https://github.com/asymmetric/nixpkgs-update-notifier) is a matrix
bot that scrapes the [nixpkgs-update logs](https://nixpkgs-update-logs.nix-community.org/) and notifies
you if nixpkgs-update/@r-ryantm fails to update any of the packages you've subscribed to.

[repology.org](https://repology.org) tracks and compares the versions of packages between various package repositories, letting you know what packages may be out of date or insecure.
You can view what packages a specific maintainer maintains and subscribe to updates with atom/rss. Example: [repology.org/maintainer/pbsds](https://repology.org/maintainer/pbsds%40hotmail.com).

> [nixpk.gs/pr-tracker](https://nixpk.gs/pr-tracker.html) and [nixpkgs-tracker.ocfox.me](https://nixpkgs-tracker.ocfox.me/)

Reviewer 3: I would then also mention this tool: https://nixpkgs-tracker.ocfox.me/ (I personally prefer it)

Author: already mentioned. i made sure to include it since it is not limited by the api restrictions of the gh token on the server

Reviewer 3: Oops, sorry !

Reviewer: Neither is pr-tracker in practice. The rate limit is 5000 requests per hour, which I would be shocked if it's ever been hit. If something goes wrong with pr-tracker it tends to be git garbage collection having failed and disk space having run out as a result.

Reviewer 6: No problem. 🥰

The maintainer is welcome to come back at any time.

## Tools for maintainers

When a pull request is made against a package, nixpkgs CI will notify the appropriate
maintainer(s) by trying to correlate the files the PR touches with the packages that need rebuilding.

Maintainers can also invoke the [nixpkgs-merge-bot](https://github.com/nixos/nixpkgs-merge-bot)
to merge pull requests targeting packages they are the maintainer of which satisfy the current
security [constraints](https://github.com/NixOS/nixpkgs-merge-bot/blob/main/README.md#constraints).
Examples: [#397273](https://github.com/NixOS/nixpkgs/pull/397273#issuecomment-2789382120) and [#377027](https://github.com/NixOS/nixpkgs/pull/377027#issuecomment-2614510869)

New maintainers will automatically get invited to join the [NixOS/nixpkgs-maintainers](https://github.com/orgs/NixOS/teams/nixpkgs-maintainers) GitHub team.

> By joining it you will get some triaging rights in the nixpkgs repository, like the ability to add and remove labels.

Author: ```suggestion
By joining it you will get some triaging rights in the nixpkgs repository, like the ability to close and reopen issues made by others, and managing labels.
```

> However, the GitHub team invite is usually only sent by email and is only valid for one week!

Reviewer 7: ```suggestion
However, the GitHub team invite is usually only sent by email, and is only valid for one week!
```

> Should it expire please ask for a re-invite in the [GitHub org owners help desk Matrix channel](https://matrix.to/#/#org_owners:nixos.org).

Reviewer 7: ```suggestion
Should it expire, please ask for a re-invite in the [GitHub org owners help desk Matrix channel](https://matrix.to/#/#org_owners:nixos.org).
```

> #### Unofficial maintainer tooling

Reviewer 2: https://github.com/Defelo/nixpkgs-review-gha is also worth mentioning, it lets you run review on architectures you don't own

Author: This is more in the scope of review tooling however, not maintenance tooling. I'd be up for adding it but I think it makes sense for a follow up PR :+1:

Reviewer: I remember us being nervous about pointing people to it on some previous occasion, due to uncertainty about how GitHub would feel about this use of Actions.

> #### Unofficial maintainer tooling

Author: I did this out of habit, being used to subsections on level 4 and below to not be automatically numbered :sweat_smile:

The maintainer is welcome to come back at any time.

## Tools for maintainers

When a pull request is made against a package, nixpkgs CI will notify the
appropriate maintainer(s) based on the package names listed in the commit titles.

> Maintainers can also invoke the [nixpkgs-merge-bot](https://github.com/nixos/nixpkgs-merge-bot)

Reviewer 7: Could we link to an "example PR", where someone used `nixpkgs-merge-bot` to merge a PR? With consent from the demonstrator, of course.

Reviewer 8: Feel free to link to #397273 (created by @r-ryantm) and/or (if it's fine for @alyssais) to #377027 (created by a committer)

Reviewer 7: Perfect, thank you! Feel free to mark this as resolved.

to merge pull requests targeting packages they are the maintainer of which satisfy the current
security [constraints](https://github.com/NixOS/nixpkgs-merge-bot/blob/main/README.md#constraints).

New maintainers will automatically get invited to join the [NixOS/nixpkgs-maintainers](https://github.com/orgs/NixOS/teams/nixpkgs-maintainers) GitHub team.
By joining it you will get some triaging rights in the nixpkgs repository, like the ability to add and remove labels.
However, the GitHub team invite is usually only sent by email and is only valid for one week!
Should it expire please ask for a re-invite in the [GitHub org owners help desk Matrix channel](https://matrix.to/#/#org_owners:nixos.org).

[zh.fail](https://zh.fail/failed/overview.html) tracks all package build failures on `master` grouped by maintainer.
Alternatively the nixpkgs CI [hydra.nixos.org](https://hydra.nixos.org/) allows you to star any job whose status you care about.

> This results in a dashboard with an overview of all starred jobs. Example: [hydra.nixos.org/dashboard/pbsds](https://hydra.nixos.org/dashboard/pbsds@hotmail.com)

Reviewer 9: Yeah, I think the way you are using it is a bit excessive, and we should build a better platform for that.

Author: zh.fail tracking both `master` and `release-xx.yy` could be a solution. In which repository is it developed?

Author: The "Next check" badge reveals https://git.helsinki.tools/janne.hess/zhf. 

(off-topic) @dasJ do you think zh.fail could track multiple jobsets like nix-review-tools-reports does? Would it also make sense to migrate the repo to the nixos github org or do you prefer to keep it where it currently is?
